File: 067.AddBinary/add.py
class Solution(object):
    def addBinary(self, a, b):
        """
        :type a: str
        :type b: str
        :rtype: str
        """
        ia = len(a) - 1
        ib = len(b) - 1
        carry = 0
        res = ""
        while ia > 0 or ib > 0 or carry:
            va = 0
            vb = 0
            if ia >= 0:
                va = int(a[ia])
            if ib >= 0:
                vb = int(b[ib])

            sum = va + vb + carry
            res += str(sum % 2)
            carry = sum / 2

            ia -= 1
            ib -= 1

        return res[::-1]

if __name__ == '__main__':
    print Solution().addBinary("11", "1")

File: 067.AddBinary/add.cc
#include <iostream>
#include <string>
#include <algorithm>

using namespace std;

class Solution {
public:
    string addBinary(string a, string b) {
        int ia = a.size()-1, ib = b.size()-1;
        int carry = 0;
        string res = "";
        while (ia >= 0 || ib >= 0 || carry >0) {
            int va = 0, vb = 0;
            if (ia >= 0) {
                va = a[ia--] - '0';
            }
            if (ib >= 0) {
                vb = b[ib--] - '0';
            }
            int sum = va + vb + carry;

            res.push_back(char(sum % 2 + '0'));
            carry = sum / 2;
        }

        reverse(res.begin(), res.end());
        return res;
    }
};

int main(int argc, char const *argv[]) {
    Solution s = Solution();
    cout << s.addBinary("11", "1") << endl;
    return 0;
}
